glean_parser/python_server.py

```diff
@@ -0,0 +1,130 @@
+    return f"create_{util.snake_case(ping)}_server_event_logger"
+
+
+def generate_event_record_function_name(event_metric: metrics.Metric) -> str:
+    return (
+        f"record_{util.snake_case(event_metric.category)}_"
+        + f"{util.snake_case(event_metric.name)}"
+    )
+
+
+def output_python(
+    objs: metrics.ObjectTree, output_dir: Path, options: Optional[Dict[str, Any]]
+) -> None:
+    """
+    Given a tree of objects, output Python code to `output_dir`.
+
+    The output is a file containing all the code for assembling pings with
+    metrics, serializing, and submitting, and an empty `__init__.py` file to
+    make the directory a package.
+
+    :param objects: A tree of objects (metrics and pings) as returned from
+        `parser.parse_objects`.
+    :param output_dir: Path to an output directory to write to.
+    """
+
+    template = util.get_jinja2_template(
+        "python_server.jinja2",
+        filters=(
+            ("camelize", camelize),
+            ("py_metric_type", generate_metric_type),
+            ("clean_string", clean_string),
+            ("factory_method", generate_ping_factory_method),
+            ("record_event_function_name", generate_event_record_function_name),
+        ),
+    )
+
+    # Go through all metrics in objs and build a map of
+    # ping->list of metric categories->list of metrics
+    # for easier processing in the template.
+    ping_to_metrics: Dict[str, Dict[str, List[metrics.Metric]]] = defaultdict(dict)
+    for _category_key, category_val in objs.items():
+        for _metric_name, metric in category_val.items():
+            if isinstance(metric, metrics.Metric):
+                if metric.type not in SUPPORTED_METRIC_TYPES:
+                    print(
+                        "❌ Ignoring unsupported metric type: "
+                        + f"{metric.type}:{metric.name}."
+                        + " Reach out to Glean team to add support for this"
+                        + " metric type."
+                    )
+                    continue
+                for ping in metric.send_in_pings:
+                    metrics_by_type = ping_to_metrics[ping]
+                    metrics_list = metrics_by_type.setdefault(metric.type, [])
+                    metrics_list.append(metric)
+
+    for ping, metrics_by_type in ping_to_metrics.items():
+        if "event" not in metrics_by_type:
+            print(
+                f"❌ No event metrics found for ping: {ping}."
+                + " At least one event metric is required."
+            )
+            return
+
+    extension = ".py"
+    filepath = output_dir / ("server_events" + extension)
+    with filepath.open("w", encoding="utf-8") as fd:
+        fd.write(template.render(parser_version=__version__, pings=ping_to_metrics))
+
+    # create an empty `__init__.py` file to make the directory a package
+    init_file = output_dir / "__init__.py"
+    with init_file.open("w", encoding="utf-8") as fd:
```

Comment:
Suggestion: This could create it if it doesn't exist. I was using our `glean/__init__.py`, was a little surprised to see it overwritten.

Reply:
Is https://github.com/mozilla/fx-private-relay/commit/51f5e88643822b7a83e4982df0aea60bc728fb7c#diff-5d5a21ba24b718227a3e2e46a94a9333471dfd62ef7b673fb7b5d7de9e84da8eL1-L3 all that you'd need in the package init? If so, we can generate this too. I'd be inclined to keep the overwrite behavior to keep this consistent with the rest of the glean_parser.

Reply:
Yep, but there isn't much different between:

```
from privaterelay.glean import EventsServerEventLogger, create_events_server_event_logger
```

and
```
from privaterelay.glean.server_events import EventsServerEventLogger, create_events_server_event_logger
```

I don't think it is worth the complexity of generating a non-blank `__init__,py` to break consistency with the rest of `glean_parser`. Withdrawing my suggestion.